### PR TITLE
raftstore: fix incorrect cond judgment leading to rejection of pre_proposal

### LINKED ISSUE #14219: 【witness】inject one tikv failure for 10m, pitr task became to paused
### What version of TiKV are you using?

./tikv-server -V
 TiKV 
Release Version:   6.6.0-alpha
Edition:           Community
Git Commit Hash:   c9cebe6cd44255018202d2e56023df096038bc09
Git Commit Branch: heads/refs/tags/v6.7.0-alpha
UTC Build Time:    2023-02-13 11:16:01
Rust Version:      rustc 1.67.0-nightly (96ddd32c4 2022-11-14)
Enable Features:   pprof-fp jemalloc mem-profiling portable sse test-engine-kv-rocksdb test-engine-raft-raft-engine cloud-aws cloud-gcp cloud-azure
Profile:           dist_release
2023-02-14T06:23:02.448Z	INFO	k8s/client.go:132	it should be noted that a long-running command will not be interrupted even the use case has ended. For more information, please refer to https://github.com/pingcap/test-infra/discussions/129
### Steps to reproduce
https://tcms.pingcap.net/dashboard/executions/plan/1562223
clinic：https://clinic.pingcap.com.cn/portal/#/orgs/31/clusters/7199855553741822161
logs：http://minio.pingcap.net:38888/buckets/test-infra-testground/browse/YXJjaGl2ZS9lbmRsZXNzLWhhLXRlc3Qtd2l0bmVzcy10cHMtMTU2MjIyMy0xLTcyOS8=
1、br restore data
2、enable witness
3、enable pitr
4、run tpcc run
5、inject one tikv failure for 10m
6、pitr task became to paused

### What did you expect?
inject one tikv failure for 10m, pitr task should be normal

### What did happened?
inject one tikv failure for 10m, pitr task became to paused
643a525a-abfe-4ee2-9f56-27ee7a437fde

### PR DESCRIPTION
**Pull request by nathan-edgar**
close tikv/tikv#14219

### What is changed and how it works?

Issue Number: Close #14219 

What's Changed:

Removed incorrect conditional judgment. The triggering scenario is:

1. The Voter has received the switch witness request, but has not yet applied it;
2. Raft elections occur, and the Voter is elected to become the leader;
3. After the Voter becomes the leader, the switch witness is applied to become the witness;

This incorrect condition judgment will incorrectly set `wait_data` to `true`, resulting in the incorrect rejection of the transfer leader request sent by PD in the pre_proposal stage, and eventually causing PITR to time out.

The left changes are:
1.  To avoid duplication of request snapshots, which has nothing to do with the scenario described above and was added for convenience;
2. Notify pd immediately, when the leader learns that non-witness has applied the snapshot;
3. Change  BatchSwitchWitness as urgent request;
 

```commit-message
raftstore: fix incorrect cond judgment leading to rejection of pre_proposal
```

### Related changes

- None

### Check List 

Tests 

- E2E test: https://tcms.pingcap.net/dashboard/executions/plan/1680594 (The failure of other cases is that the region is unbalanced for a long time, which has nothing to do with this repair)

Side effects

- Performance regression
    - None
    
### Release note 

```release-note
None
```

### CODE REVIEW

**Review by Connor1996 — COMMENTED** [review]
unit test?

File: components/raftstore/src/store/fsm/peer.rs [hunk 1]
```diff
@@ -6455,7 +6466,7 @@ where
         for s in sw.switches {
             let (peer_id, is_witness) = (s.get_peer_id(), s.get_is_witness());
             if self.fsm.peer_id() == peer_id {
-                if is_witness && !self.fsm.peer.is_leader() {
```
[the remark this answers is not in the document]
Comment (nathan-edgar) [reply components/raftstore/src/store/fsm/peer.rs:6458]: Added a `delay_clean_data` flag, will be cleaned up in `on_role_changed`, thanks! (Will add some tests to cover)

File: components/raftstore/src/store/fsm/peer.rs [hunk 2]
```diff
@@ -2660,6 +2660,13 @@ where
                 "receive peer ready info";
                 "peer_id" => self.fsm.peer.peer.get_id(),
             );
+            info!(
```
Comment (BusyJay) [inline components/raftstore/src/store/fsm/peer.rs:2663]: Better check if `self.fsm.peer.wait_data_peers` is really updated.

**Review by BusyJay — APPROVED** [review]
Rest LGTM

**Review by Connor1996 — APPROVED** [review]
LGTM

**Pull request merged**